pymc/logprob/order.py

```diff
@@ -0,0 +1,216 @@
+#
+#   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+#   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+#   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+#   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+#   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+#   SOFTWARE.
+
+from typing import List, Optional
+
+import pytensor.tensor as pt
+
+from pytensor.graph.basic import Node
+from pytensor.graph.fg import FunctionGraph
+from pytensor.graph.rewriting.basic import node_rewriter
+from pytensor.tensor.math import Max
+from pytensor.tensor.random.op import RandomVariable
+from pytensor.tensor.var import TensorVariable
+
+from pymc.logprob.abstract import (
+    MeasurableVariable,
+    _logcdf_helper,
+    _logprob,
+    _logprob_helper,
+)
+from pymc.logprob.rewriting import measurable_ir_rewrites_db
+
+
+class MeasurableMax(Max):
+    """A placeholder used to specify a log-likelihood for a cmax sub-graph."""
+
+
+MeasurableVariable.register(MeasurableMax)
+
+
+@node_rewriter([Max])
+def find_measurable_max(fgraph: FunctionGraph, node: Node) -> Optional[List[TensorVariable]]:
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+    if rv_map_feature is None:
+        return None  # pragma: no cover
+
+    if isinstance(node.op, MeasurableMax):
+        return None  # pragma: no cover
+
+    base_var = node.inputs[0]
+
+    if base_var.owner is None:
+        return None
+
+    # NonRVS must be rejected
+    if not isinstance(base_var.owner.op, RandomVariable):
+        return None
+
+    # univariate iid test which also rules out other distributions
+    if isinstance(base_var.owner.op, RandomVariable):
+        for params in base_var.owner.inputs[3:]:
+            if params.type.ndim != 0:
+                return None
+
+    if not rv_map_feature.request_measurable(node.inputs):
+        return None
+
+    # Check whether axis is supported or not
+    axis = set(node.op.axis)
+    base_var_dims = set(range(base_var.ndim))
+    if axis != base_var_dims:
+        return None
+
+    measurable_max = MeasurableMax(list(axis))
+    max_rv_node = measurable_max.make_node(base_var)
+    max_rv = max_rv_node.outputs
+
+    return max_rv
+
+
+measurable_ir_rewrites_db.register(
+    "find_measurable_max",
+    find_measurable_max,
+    "basic",
+    "max",
+)
+
+
+@_logprob.register(MeasurableMax)
+def max_logprob(op, values, base_rv, **kwargs):
+    r"""Compute the log-likelihood graph for the `Max` operation.
+
+    Parameters
+    ----------
+    op : Max-Op
+    values : tensor_like
+    rv : TensorVariable
+
+    Returns
+    -------
+    logprob : TensorVariable
+
+    Examples
+    --------
+    It is often desirable to find the Maximum from the distribution of random variables.
```

Comment:
```suggestion
    It is often desirable to find the log-probability of the maximum of i.i.d. random variables.
```
I think that the delineation between finding the maximum vs. finding the log-probability of the maximum of random variables are different tasks, the latter of which is your focus (and the more challenging one of the two)!

Reply:
Yes I completely agree. What I intended to do here is first give a gist of what a maximum operation is and how users can use this themselves before introducing them to finding logp of the maximum of random variables. Maybe I can rephrase this in a better way.